feat(ot3): ot3 hardware controller move

amitlissack opened the pull request:
# Overview

It's a start.

# Changelog

- adding `enableOT3HardwareController` feature flag.
- use feature flag to use `OT3Controller`.
- ot3 controller connects to can bus
- move command implemented.

# Review requests



# Risk assessment

HIGH. This cannot be merged until we validate that it doesn't interfere with OT2 operation.

Commit:
Ot3 controller fills out move

deal with module not found for the ot3 stuff.

pipfile.

sfoster1 commented on api/Pipfile:
```diff
@@ -36,3 +36,4 @@ types-mock = "==4.0.1"
 types-setuptools = "==57.0.2"
 opentrons-shared-data = {editable = true, path = "../shared-data/python"}
 opentrons = {editable = true, path = "."}
+opentrons-hardware = {editable = true, path = "./../hardware"}
```
i _think_ the right way to do this is actually to put this in the `setup.py` as an `extras_require`. then in this pipfile we can install the extra, by making line 38 `opentrons = {editable = true, path = ".", extras=["hardware"]}` or whatever.

amitlissack replied:
Would this require that `opentrons_hardware` be in pypy?

sfoster1 replied:
oh. uhhhhh yeah it might that would probably be annoying

sfoster1 commented on robot-server/Pipfile:
```diff
@@ -48,3 +48,4 @@ idna = "==3.2"
 click = "==8.0.1"
 numpy = "==1.15.1"
 zipp = "==3.5.0"
+opentrons-hardware = {editable = true, path = "./../hardware"}
```
i think we won't need this when we set the hardware package as an optional dep of `opentrons`, particularly if we have it marked as an extra in the robot server setup.py

sfoster1 reviewed (approved):
I think we can slightly change the way we set the dependencies to be a little smoother but the actual functionality looks great!

sfoster1 reviewed (commented):
What if the runtime stuff, rather than (or in addition to) being gated by the feature flag, was gated by whether `hardware` was present and importable? then it wouldn't be possible to accidentally enable it

Commit:
ot3 hardware controller can only be used if opentrons_hardware package is available.

Commit:
change default can network to can0

Commit:
fixes

Commit:
more fixes.

Commit:
logging.

Commit:
start the messenger.

Commit:
bug

Commit:
setup and teardown motors.

sfoster1 commented on hardware/opentrons_hardware/hardware_control/move_group_runner.py:
```diff
@@ -43,9 +46,17 @@ async def run(self, can_messenger: CanMessenger) -> None:
+        try:
+            await self._move(can_messenger)
+        finally:
+            await self._teardown_motors(can_messenger)
```
We probably don't want to systematically do this; in general, we want to leave motors enabled unless explicitly told not to because without encoders we'll lose positioning when we turn the motors off, and even with encoders we'll fuzz our position to encoder resolution.

amitlissack replied:
Okeedoke

sfoster1 commented on hardware/opentrons_hardware/hardware_control/move_group_runner.py:
```diff
@@ -43,9 +46,17 @@ async def run(self, can_messenger: CanMessenger) -> None:
         await self._clear_groups(can_messenger)
         await self._send_groups(can_messenger)
-        await self._move(can_messenger)
+        await self._setup_motors(can_messenger)
```
we should put a todo here to remove this when possible - we should probably make the firmware handle this on its own

sfoster1 reviewed (approved):
Looks good! Couple small comments but really exciting

Commit:
add py.typed to opentrons_hardware.

Commit:
lint

Commit:
black.

Commit:
specify can channel in ci.

Commit:
don't engage and disengage motors between moves.

sfoster1 commented on api/src/opentrons/__init__.py:
```diff
@@ -106,12 +119,18 @@ async def _create_thread_manager() -> ThreadManager:
-        port=_get_motor_control_serial_port(),
-        firmware=_find_smoothie_file(),
-    )
+    if should_use_ot3():
```
after thinking about it more - maybe we should do this switch in the robot server? it does seem like something that the robot server would care about, and we could then locate the dependency on `opentrons_hardware` there

Commit:
fix test.

sfoster1 committed:
feat(hardware): support two head axes (#8946)

Add the head_l and head_r nodeids; refactor the move group creator to
iterate over provided axes rather than naming them explicitly and
therefore be able to handle the new node ids.
